### PR TITLE
Get device info for SSO ext. Operation [common core]

### PR DESCRIPTION
**Pull request by jasoncoolmax**
This is a branch which cherry picks all the changes from [this one](https://github.com/AzureAD/microsoft-authentication-library-common-for-objc/pull/603) and points to the right target branch.

### CODE REVIEW

File: IdentityCore/src/broker_operation/response/MSIDBrokerOperationResponse.m [hunk 1]
```diff
@@ -33,17 +38,26 @@ - (instancetype)initWithJSONDictionary:(NSDictionary *)json error:(NSError **)er
...
 }
 
 - (NSDictionary *)jsonDictionary
 {
-    // TODO: implement.
+    NSMutableDictionary *json = [NSMutableDictionary new];
+    json[MSID_BROKER_OPERATION_JSON_KEY] = self.operation;
+    json[MSID_BROKER_OPERATION_RESULT_JSON_KEY] = [@(self.success) stringValue];
+    
+    json[@"device_info"] = self.deviceInfo.jsonDictionary;
```
Comment (oldalton) [inline IdentityCore/src/broker_operation/response/MSIDBrokerOperationResponse.m]: this creates a nested dictionary, but initializer doesn't expect a nested dictionary.

File: IdentityCore/src/broker_operation/response/MSIDDeviceInfo.h [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#import "MSIDJsonSerializable.h"
+
+typedef NS_ENUM(NSInteger, MSIDDeviceShareMode)
```
Comment (oldalton) [inline IdentityCore/src/broker_operation/response/MSIDDeviceInfo.h]: let's just call it deviceMode such as in future we could have modes not related to shared devices?
We could start with MSIDDeviceModePersonal and MSIDDeviceModeShared to start with.

File: IdentityCore/src/broker_operation/response/MSIDDeviceInfo.m [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+#pragma mark - MSIDJsonSerializable
+
+- (instancetype)initWithJSONDictionary:(NSDictionary *)json error:(NSError **)error
+{
+    self = [super init];
+    
+    if (self)
+    {
+        _deviceMode = [[json msidStringObjectForKey:MSID_BROKER_DEVICE_MODE_KEY] intValue];
```
Comment (oldalton) [inline IdentityCore/src/broker_operation/response/MSIDDeviceInfo.m]: should these be passed as strings for readability of the protocol if someone wants to use it without MSAL?

File: IdentityCore/src/broker_operation/response/MSIDBrokerOperationResponse.m (hunk 1 above)
Comment (jasoncoolmax) [reply IdentityCore/src/broker_operation/response/MSIDBrokerOperationResponse.m]: addressed :)

File: IdentityCore/src/broker_operation/response/MSIDDeviceInfo.h (hunk 2 above)
Comment (jasoncoolmax) [reply IdentityCore/src/broker_operation/response/MSIDDeviceInfo.h]: sure, addressed :)

File: IdentityCore/src/broker_operation/response/MSIDDeviceInfo.m (hunk 3 above)
Comment (jasoncoolmax) [reply IdentityCore/src/broker_operation/response/MSIDDeviceInfo.m]: addressed :)

File: IdentityCore/src/broker_operation/response/MSIDDeviceInfo.h [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+};
+#define WPJStatusStringArray @"NotJoined", @"joined", nil
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface MSIDDeviceInfo : NSObject <MSIDJsonSerializable>
+
+@property (nonatomic) MSIDDeviceMode deviceMode;
+@property (nonatomic) MSIDWorkPlaceJoinStatus wpjStatus;
+@property (nonatomic, nullable) NSString *brokerVersion;
```
Comment (antrix1989) [inline IdentityCore/src/broker_operation/response/MSIDDeviceInfo.h]: This is broker sdk version, not authenticator app, correct?

File: IdentityCore/src/broker_operation/response/MSIDDeviceInfo.m [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+
+- (instancetype)initWithJSONDictionary:(NSDictionary *)json error:(__unused NSError **)error
+{
+    self = [super init];
+    
+    if (self)
+    {
+        _deviceMode = [self deviceModeEnumFromString:[json msidStringObjectForKey:MSID_BROKER_DEVICE_MODE_KEY]];
+        _wpjStatus = [self wpjStatusEnumFromString:[json msidStringObjectForKey:MSID_BROKER_WPJ_STATUS_KEY]];
+        _brokerVersion = [json msidStringObjectForKey:MSID_BROKER_BROKER_VERSION_KEY];
```
Comment (antrix1989) [inline IdentityCore/src/broker_operation/response/MSIDDeviceInfo.m]: nit: BROKER_BROKER

File: IdentityCore/src/broker_operation/response/MSIDDeviceInfo.m [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+    NSUInteger index = [self.deviceModeEnumString indexOfObject:deviceModeString];
+    
+    if (index < 0 || index >= self.deviceModeEnumString.count)
+    {
+        return (MSIDDeviceMode) 0;
+    }
+    return (MSIDDeviceMode) index;
+}
+
+- (NSArray *)deviceModeEnumString
```
Comment (antrix1989) [inline IdentityCore/src/broker_operation/response/MSIDDeviceInfo.m]: Should we rather use "switch" were we convert enum to string and vice versa? Seem to be less code and no need in "#define" ...

File: IdentityCore/src/broker_operation/response/MSIDDeviceInfo.h (hunk 4 above)
Comment (jasoncoolmax) [reply IdentityCore/src/broker_operation/response/MSIDDeviceInfo.h]: yes, it is broker SDK version :)

File: IdentityCore/src/broker_operation/response/MSIDDeviceInfo.m (hunk 5 above)
Comment (jasoncoolmax) [reply IdentityCore/src/broker_operation/response/MSIDDeviceInfo.m]: As shown in this [file](https://github.com/AzureAD/microsoft-authentication-library-common-for-objc/blob/dev/IdentityCore/src/MSIDBrokerConstants.h), `MSID_BROKER` is the prefix` for those constants. And `BROKER_VERSION` is the name of the constant here :)

File: IdentityCore/src/broker_operation/response/MSIDDeviceInfo.m (hunk 6 above)
Comment (jasoncoolmax) [reply IdentityCore/src/broker_operation/response/MSIDDeviceInfo.m]: Agree, changed :)

File: IdentityCore/src/broker_operation/response/MSIDDeviceInfo.m [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+    if ([deviceModeString isEqualToString:@"shared"])  return MSIDDeviceModeShared;
+
+    return MSIDDeviceModePersonal;
+}
+
+- (NSString *)wpjStatusStringFromEnum:(MSIDWorkPlaceJoinStatus)wpjStatus
+{
+    switch (wpjStatus) {
+        case MSIDWorkPlaceJoinStatusNotJoined:
+            return @"NotJoined";
```
Comment (oldalton) [inline IdentityCore/src/broker_operation/response/MSIDDeviceInfo.m]: nit: Should we have consistent casing between "NotJoined" and "joined" :)?

Comment (jasoncoolmax) [reply IdentityCore/src/broker_operation/response/MSIDDeviceInfo.m]: addressed :)

**Pull request merged**